### ISSUE TITLE
DELETE request doesn't decode payload 

### ISSUE THREAD

**Issue by yamalight** [fastify/fastify#485]
Currently if you try to send DELETE request that has payload, fastify won't pass the decoded payload to handler. E.g.:
```js
fastify.route({
  method: 'DELETE',
  path: '/test',
  handler(request, reply) {
    console.log(request.body); // null
  },
});

// test injection
server.inject({
  method: 'DELETE',
  url: '/test',
  payload: {
    tokenName: 'test',
  },
}, () => {});
```

Is that intended or a bug?

**Comment by StarpTech** [comment]
Hi @yamalight this intentional. DELETE, GET and HEAD don't include payload. For more information look at https://tools.ietf.org/html/rfc7231#section-4.3.5

**Comment by jsumners** [comment]
From your link @StarpTech :

> A payload within a DELETE request message has no defined semantics;
   sending a payload body on a DELETE request might cause some existing
   implementations to reject the request.

Thus, it isn't **prohibited** but it also isn't defined what should happen. This means some clients won't bother sending a body even if you specify one. But that's not what is happening. @yamalight is using all Fastify internals to demonstrate the behavior.

**Comment by yamalight** [comment]
It's definitely not prohibited by spec. Other frameworks support this behavior (e.g. hapi, express w/ body-parser)

**Comment by StarpTech** [comment]
So it's a framework decision. When most of the Node.js framework support payload in DELETE requests, we should support it too.

**Comment by StarpTech** [comment]
@yamalight could you prepare a PR ?

**Comment by StarpTech** [comment]
Hapi supports it https://github.com/hapijs/hapi/blob/3b9ec87249845e4ebab0f4769fb446d82a20c608/lib/route.js#L158
Express https://github.com/expressjs/body-parser/blob/3cb380b1e4b3cabdfb7db265dd7a0dd637b7687a/test/body-parser.js#L75
Koa support it too but provide a `strict` option https://github.com/dlau/koa-body#options

**Comment by yamalight** [comment]
@StarpTech sure, I can try to :)

**Comment by yamalight** [comment]
@StarpTech OK, so I've built a simple version here: https://github.com/yamalight/fastify/commit/5e079cea18f4000b65e6a31e20c637070de88536
But there's a small problem that I cannot solve - the object sent back from that new `.delete` handler is never valid JSON, so my new test fails. Any ideas why that might happen? Have I missed something?

**Comment by delvedor** [comment]
@yamalight add the `DELETE` check [here](https://github.com/fastify/fastify/blob/master/lib/handleRequest.js#L31).

Can you open the pr? In that way we can see the actual error.

**Comment by yamalight** [comment]
@delvedor PR created here: https://github.com/fastify/fastify/pull/486
Have I understood you / moved the check correctly?

**Comment by delvedor** [comment]
Closed by #486
